tests/exporters/test_vhi.py

```diff
@@ -0,0 +1,68 @@
+from pathlib import Path
+# from unittest.mock import patch, Mock
+import pytest
+
+from src.exporters.vhi import (
+    VHIExporter,
+    _parse_time_from_filename,
+    make_filename,
+    # download_file_from_ftp
+)
+
+import sys
+# get the root project directory
+project_dir = Path(__file__).resolve().parents[2]
+# append to sys
+sys.path.append(project_dir.as_posix())
+
+
+class TestVHIExporter:
+
+    def test_initialisation(self):
+        VHIExporter()
+        return
+
+    def test_filename_illegitimate(self):
+        with pytest.raises(Exception):
+            make_filename(
+                raw_folder=project_dir / "data",
+                raw_filename='this/should/fail.nc',
+                dataset='vhi',
+            )
+
+        return
+
+    def test_filenames_found(self):
```

Comment:
As we discussed, I would prefer if we didn't actually call the NOAA website to check this code works.

One way to avoid this is by patching ftplib: https://stackoverflow.com/questions/29817963/mocking-ftplib-ftp-for-unit-testing-python-code

Reply:
My test function

```
    @patch('ftplib.FTP', autospec=True)
    def test_filenames_found(self, mock_ftp_constructor):
        mock_ftp = mock_ftp_constructor.return_value
        vhi = VHIExporter()
        fnames = vhi.get_ftp_filenames()

        mock_ftp_constructor.assert_called_with('ftp.star.nesdis.noaa.gov')
        self.assertTrue(mock_ftp.login.called)

```

My real function:

```
    @staticmethod
    def get_ftp_filenames() -> List:
        """  get the filenames containing VHI """
        with FTP('ftp.star.nesdis.noaa.gov') as ftp:
            ftp.login()
            ftp.cwd('/pub/corp/scsb/wguo/data/Blended_VH_4km/VH/')

            # append the filenames to a list
            listing: List = []
            ftp.retrlines("LIST", listing.append)
            # extract the filename
            filepaths = [f.split(' ')[-1] for f in listing]
            # extract only the filenames of interest
            vhi_files = [f for f in filepaths if ".VH.nc" in f]

        return vhi_files

```

Do I need to change the function to separate teh

Reply:
Error message

```
tests/exporters/test_vhi.py:38:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

_mock_self = <MagicMock name='FTP' spec='FTP' id='4626376352'>, args = ('ftp.star.nesdis.noaa.gov',)
kwargs = {}, expected = "FTP('ftp.star.nesdis.noaa.gov')"

    def assert_called_with(_mock_self, *args, **kwargs):
        """assert that the mock was called with the specified arguments.

        Raises an AssertionError if the args and keyword args passed in are
        different to the last call to the mock."""
        self = _mock_self
        if self.call_args is None:
            expected = self._format_mock_call_signature(args, kwargs)
>           raise AssertionError('Expected call: %s\nNot called' % (expected,))
E           AssertionError: Expected call: FTP('ftp.star.nesdis.noaa.gov')
E           Not called

../../miniconda3/envs/crop/lib/python3.7/unittest/mock.py:811: AssertionError
```

Reply:
I don't think it needs to be split into two functions. So one thing you could do is mock the `ftp.retrlines` function so that it returns a bunch of filenames (since what we want to test is that we only retrieve the `.VH.nc` filenames).

Mock has a `side_effect` feature (see this [blog post](http://fgimian.github.io/blog/2014/04/10/using-the-python-mock-library-to-fake-regular-functions-during-tests/)) which allows you to replace a function with another, so I would do that for `ftp.retrlines`, something like this:

```python
def fake_retrlines(some_string, callback):
    # I don't know if the string argument is meaningful; if it is it can be considered
    # here too
    fake_data = ['a', 'b', 'c.VI.nc', 'd', 'e.VI.nc']
    for filename in fake_data:
        callback(filename)
```

And then, within the function, add:
`@mock.patch('ftp.retrlines', side_effect=fake_retrlines)`

Reply:
I don't think it needs to be split into two functions. So one thing you could do is mock the ftp.retrlines function so that it returns a bunch of filenames (since what we want to test is that we only retrieve the .VH.nc filenames).

Mock has a side_effect feature (see this blog post) which allows you to replace a function with another, so I would do that for ftp.retrlines, something like this:

```python
def fake_retrlines(some_string, callback):
    # I don't know if the string argument is meaningful; if it is it can be considered
    # here too
    fake_data = ['a', 'b', 'c.VI.nc', 'd', 'e.VI.nc']
    for filename in fake_data:
        callback(filename)
```
And then, within the function, add: `@mock.patch('ftp.retrlines', side_effect=fake_retrlines)`

Reply:
https://stackoverflow.com/questions/28850070/python-mocking-a-context-manager?noredirect=1&lq=1

Reply:
https://stackoverflow.com/questions/29817963/mocking-ftplib-ftp-for-unit-testing-python-code

Reply:
Attempt 2:
```
    ```@patch('ftplib.FTP', autospec=True)
    def test_filenames_found(self, mock_ftp):
        mock_ftp.return_value = Mock()
        mock_ftp_obj = mock_ftp()

        fnames = VHIExporter.get_ftp_filenames()
        mock_ftp.assert_called_with('ftp.star.nesdis.noaa.gov')```
```

Reply:
attempt 3:
```
    ```def test_filenames_found(self, mock_ftp):
        mock_ftp.return_value.__enter__.return_value.name = Mock()
        mock_ftp_constructor = Mock()
        vhi = VHIExporter()
        fnames = vhi.get_ftp_filenames()

        mock_ftp.assert_called_once_with('ftp.star.nesdis.noaa.gov')```
```

Reply:
@gabrieltseng's answer:
```
    @patch('ftplib.FTP', autospec=True)
    def test_ftp_called(self, mock_ftp):
        _ = VHIExporter.get_ftp_filenames(years=list(range(1981, 2020)))
        mock_ftp.assert_called_once_with('ftp.star.nesdis.noaa.gov')

```